zookeeper-server/src/main/java/org/apache/zookeeper/server/ZKDatabase.java

```diff
@@ -187,18 +190,21 @@ public ReentrantReadWriteLock getLogLock() {
     }
 
 
-    public synchronized List<Proposal> getCommittedLog() {
+    public synchronized Collection<Proposal> getCommittedLog() {
+        final Collection<Proposal> result;
         ReadLock rl = logLock.readLock();
-        // only make a copy if this thread isn't already holding a lock
-        if(logLock.getReadHoldCount() <=0) {
+        // make a copy if this thread is not already holding a lock
+        if (logLock.getReadHoldCount() > 0) {
+          result = this.committedLog;
+        } else {
+            rl.lock();
             try {
-                rl.lock();
-                return new LinkedList<Proposal>(this.committedLog);
+                result = new ArrayList<>(this.committedLog);
             } finally {
                 rl.unlock();
             }
         }
-        return this.committedLog;
+        return Collections.unmodifiableCollection(result);
```

Comment:
For other reviewers - the changes here are the most interesting (imo).

Reply:
Thanks @enixon !

In one case, the caller is getting a point-in-time copy of the committedLog queue, and in the other case, the caller is getting the real thing.  That's already a bit unexpected, but I figured at the very least, the caller should not be allowed to modify the real thing outside the scope of this class.